// revolt.js/lib/cjs/collections/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserCollection = exports.SessionCollection = exports.ServerMemberCollection = exports.ServerCollection = exports.MessageCollection = exports.EmojiCollection = exports.ChannelWebhookCollection = exports.ChannelUnreadCollection = exports.ChannelCollection = exports.BotCollection = void 0;
__exportStar(require("./Collection.js"), exports);
var BotCollection_js_1 = require("./BotCollection.js");
Object.defineProperty(exports, "BotCollection", { enumerable: true, get: function () { return BotCollection_js_1.BotCollection; } });
var ChannelCollection_js_1 = require("./ChannelCollection.js");
Object.defineProperty(exports, "ChannelCollection", { enumerable: true, get: function () { return ChannelCollection_js_1.ChannelCollection; } });
var ChannelUnreadCollection_js_1 = require("./ChannelUnreadCollection.js");
Object.defineProperty(exports, "ChannelUnreadCollection", { enumerable: true, get: function () { return ChannelUnreadCollection_js_1.ChannelUnreadCollection; } });
var ChannelWebhookCollection_js_1 = require("./ChannelWebhookCollection.js");
Object.defineProperty(exports, "ChannelWebhookCollection", { enumerable: true, get: function () { return ChannelWebhookCollection_js_1.ChannelWebhookCollection; } });
var EmojiCollection_js_1 = require("./EmojiCollection.js");
Object.defineProperty(exports, "EmojiCollection", { enumerable: true, get: function () { return EmojiCollection_js_1.EmojiCollection; } });
var MessageCollection_js_1 = require("./MessageCollection.js");
Object.defineProperty(exports, "MessageCollection", { enumerable: true, get: function () { return MessageCollection_js_1.MessageCollection; } });
var ServerCollection_js_1 = require("./ServerCollection.js");
Object.defineProperty(exports, "ServerCollection", { enumerable: true, get: function () { return ServerCollection_js_1.ServerCollection; } });
var ServerMemberCollection_js_1 = require("./ServerMemberCollection.js");
Object.defineProperty(exports, "ServerMemberCollection", { enumerable: true, get: function () { return ServerMemberCollection_js_1.ServerMemberCollection; } });
var SessionCollection_js_1 = require("./SessionCollection.js");
Object.defineProperty(exports, "SessionCollection", { enumerable: true, get: function () { return SessionCollection_js_1.SessionCollection; } });
var UserCollection_js_1 = require("./UserCollection.js");
Object.defineProperty(exports, "UserCollection", { enumerable: true, get: function () { return UserCollection_js_1.UserCollection; } });
